language-support/codegen-common/src/main/scala/com/digitalasset/daml/lf/codegen/dependencygraph/DependencyGraph.scala

```diff
@@ -5,39 +5,36 @@ package com.daml.lf.codegen.dependencygraph
 
 import com.daml.lf.codegen.Util
 import com.daml.lf.data.ImmArray.ImmArraySeq
-import com.daml.lf.codegen.lf.DefTemplateWithRecord
 import com.daml.lf.data.Ref.Identifier
-import com.daml.lf.iface.{InterfaceType, DefDataType}
-import scalaz.std.either._
+import com.daml.lf.iface.InterfaceType
 import scalaz.std.list._
 import scalaz.syntax.bifoldable._
-import scalaz.syntax.bifunctor._
 import scalaz.syntax.foldable._
 
 private[codegen] object DependencyGraph {
   def orderedDependencies(
       decls: Map[Identifier, InterfaceType]
-  ): OrderedDependencies[Identifier, Either[DefTemplateWithRecord, DefDataType.FWT]] = {
+  ): OrderedDependencies[Identifier, InterfaceType] = {
     // invariant: no type decl name equals any template alias
     val typeDeclNodes =
-      decls.to(ImmArraySeq).collect { case (qualName, InterfaceType.Normal(typeDecl)) =>
+      decls.to(ImmArraySeq).collect { case (qualName, normal: InterfaceType.Normal) =>
         (
           qualName,
           Node(
-            Right(typeDecl),
-            typeDecl.bifoldMap(Util.genTypeTopLevelDeclNames)(Util.genTypeTopLevelDeclNames),
+            normal,
+            normal.`type`.bifoldMap(Util.genTypeTopLevelDeclNames)(Util.genTypeTopLevelDeclNames),
             collectDepError = false,
           ),
         )
       }
     val templateNodes =
-      decls.to(ImmArraySeq).collect { case (qualName, InterfaceType.Template(typ, tpl)) =>
-        val recDeps = typ.foldMap(Util.genTypeTopLevelDeclNames)
-        val choiceAndKeyDeps = tpl.foldMap(Util.genTypeTopLevelDeclNames)
+      decls.to(ImmArraySeq).collect { case (qualName, template: InterfaceType.Template) =>
+        val recDeps = template.rec.foldMap(Util.genTypeTopLevelDeclNames)
+        val choiceAndKeyDeps = template.template.foldMap(Util.genTypeTopLevelDeclNames)
```

Comment:
I would prefer not to do this in such cases. When the mandate of some code is "don't forget any of the data", it's a lot easier to prevent later mistakes when the code that shouldn't forget any of the data stops compiling when you add some new data.

If on the other hand we were saying "I just want 1-2 of a bunch of fields" this would be fine.

Reply:
`template` itself is used as the item in the `Node` constructor so we need to keep it anyway, either as `template @ InterfaceType.Template(rec, templateDef)` or `template: InterfaceType.Template`, so in this case I'd keep it in.

In general I get your point but I'd prefer not break code whenever a field is added to a type. But we can save this debate for other concrete cases.